pkg/network/ovn_kubernetes.go

```diff
@@ -334,3 +381,133 @@ func listenDualStack(masterIP string) string {
 		return ""
 	}
 }
+
+// shouldUpdateDaemonSets determines if we should roll out changes to
+// the master and node daemonsets.
+//
+// On upgrades, we roll out nodes first, then masters. Downgrades, we do the
+// opposite.
+func shouldUpdateOVNK(existingNode, existingMaster *appsv1.DaemonSet, releaseVersion string) (updateNode, updateMaster bool) {
+	// Fresh cluster - full steam ahead!
+	if existingNode == nil || existingMaster == nil {
+		return true, true
+	}
+
+	nodeVersion := existingNode.GetAnnotations()["release.openshift.io/version"]
+	masterVersion := existingMaster.GetAnnotations()["release.openshift.io/version"]
+
+	// shortcut - we're all rolled out.
+	if nodeVersion == releaseVersion && masterVersion == releaseVersion {
+		return true, true
+	}
+
+	masterDirection := upgradeDirection(masterVersion, releaseVersion)
+	nodeDirection := upgradeDirection(nodeVersion, releaseVersion)
+
+	if masterDirection == directionUnknown || nodeDirection == directionUnknown {
+		log.Printf("WARNING: could not determine update direction; proceeding")
+		return true, true
+	}
+	// 9 cases
+	// +------------+---------------+-----------------+---------------+
+	// |  vs. CNO   | Master older  |   Master same   | Master newer  |
+	// +------------+---------------+-----------------+---------------+
+	// | Node older | Update node   | update node     | error         |
+	// | Node same  | wait for node | done            | update master |
+	// | Node newer | error         | wait for master | update master |
+	// +------------+---------------+-----------------+---------------+
```

Comment:
I reworded it a bit. it's naturally confusing, since none of these chars define directionality super well. Let's hope it's enough context to debug this years from now.